core/shared/src/main/scala/zio/Chunk.scala

```diff
@@ -873,6 +875,39 @@ object Chunk {
       case _: BitChunk       => ClassTag.Boolean.asInstanceOf[ClassTag[A]]
     }
 
+  /**
+   * The maximum number of elements in the buffer for fast append.
+   */
+  private val BufferSize: Int =
+    100
+
+  private final case class Add[A](start: Chunk[A], buffer: Array[AnyRef], bufferUsed: Int, chain: AtomicInteger)
```

Comment:
Maybe our `buffer` should be `ChunkBuilder` since we are trying to make that smart and efficient? We'd need a way to check `bufferUsed` and capacity, and we'd need to use `sizeHint` (and make sure it's working properly).

Reply:
Yes let me work on that.

Reply:
Ok, awesome. Please just double check `ChunkBuilder` doesn't lead to a performance regression. Wouldn't want to lose these awesome numbers!

Reply:
Will do!

Reply:
Thinking about this more I'm not sure this works. We need to be able to support actual collection operations on `AppendN` chunks which could require looking up values that are in the buffer. The `Builder` interface doesn't support that, that is the difference between the builder of a collection and a collection. We could potentially keep track of the size of the builder but if we also want to allow looking up the elements of the builder I think we are just creating another collection and at that point we might as well stick with an array.

Reply:
You're right. I mean we'd need access to the underlying array with a guarantee that `clear` isn't called... and even then, that would only work up until we specialize bit chunks.

The main thing we miss with the current approach is primitives (which we always box right now, even if unnecessary).

Reply:
We could specialize this.